### ISSUE TITLE
more crispness around `Variable` vs `Tensor`

### ISSUE THREAD

**Issue by joelgrus** [allenai/allennlp#368]
a lot of our methods are type-hinted to take `torch.Tensor` but we're really giving them `torch.autograd.Variable`s.  

in particular, this is biting me in `PytorchSeq2VecWrapper` which says that its mask is a Tensor, but it's not, and if you try to find the number of non-zero elements with

```
        num_valid = torch.sum(mask[:, 0])
```

you get another variable and you have to call `.data` on it, whereas you wouldn't if mask was really a Tensor.

(it's a nuisance that mypy doesn't pick up on this, but :man_shrugging: )

anyway, I can deal with this by doing 

```
        if isinstance(num_valid, torch.autograd.Variable):
            num_valid = num_valid.data[0]
```

but that's kind of gross. I'd like to come up with a more "global" approach for handling this and also get our type hints reflective of what we're actually doing.

**Comment by joelgrus** [comment]
duplicate of https://github.com/allenai/allennlp/issues/71
